### PR TITLE
Use rich inputs for declaring system properties

### PR DESCRIPTION
**Pull request by wolfs**

### CODE REVIEW

File: contents/caching-java-projects.adoc [hunk 1]
```diff
@@ -66,38 +66,38 @@ See also the section on dealing with <<volatile_outputs,non-repeatable outputs>>
...
 Passing absolute paths will break <<relocatability,relocatability>> of the integration test task.
-For example, with a task called `integTest`, instead of this:
 
 [source,groovy]
 ----
 include::{samplescodedir}/integration-tests/build.gradle[tags=distributionPathInput]
 ----
 
-...it is better to do this:
+There is a way out: instead of adding the absolute path directly as a system property, it is possible to add a properly
```
Comment (lptr) [inline contents/caching-java-projects.adoc]: "There is a way out" sounds too informal and a bit too strong. I'd just leave it out.

Comment (lptr) [reply contents/caching-java-projects.adoc]: Also, "properly" -> "richly", or just leave it out?

File: samples/code/integration-tests/build.gradle [hunk 2]
```diff
@@ -12,27 +12,46 @@ integTest {
...
 
 // tag::distributionPathInput[]
+// Don't do this! Breaks relocatability!
 integTest {
     systemProperty "distribution.location", file("build/dist").absolutePath
 }
 // end::distributionPathInput[]
 
 // tag::distributionDirInput[]
+class IntegrationTestDistribution implements CommandLineArgumentProvider {  // <1>
```
Comment (lptr) [inline samples/code/integration-tests/build.gradle]: I think what we want here is a `DistributionLocator` or `DistributionLocationProvider`.

File: samples/code/integration-tests/build.gradle [hunk 3]
```diff
@@ -12,27 +12,46 @@ integTest {
...
+        new IntegrationTestDistribution(distribution: file("build/dist"))   // <4>
+    )
 }
 // end::distributionDirInput[]
 
 // tag::ignoreSystemProperties[]
-integTest {
-    def ciProperties = [agentNumber: (System.getenv().get("AGENT_NUMBER") ?: 1)] // <1>
-    systemProperties(ciProperties) // <2>
+class CiEnvironment implements CommandLineArgumentProvider {
```
Comment (lptr) [inline samples/code/integration-tests/build.gradle]: Same here, it should be `CiEnvironmentProvider`.

File: samples/code/integration-tests/build.gradle (hunk 2 above)
Comment (wolfs) [reply samples/code/integration-tests/build.gradle]: Done.

File: samples/code/integration-tests/build.gradle (hunk 3 above)
Comment (wolfs) [reply samples/code/integration-tests/build.gradle]: Done.

File: contents/caching-java-projects.adoc (hunk 1 above)
Comment (wolfs) [reply contents/caching-java-projects.adoc]: Done.

**Comment by wolfs** [comment]
@lptr Please have another look.

**Pull request merged**